File: app/src/main/java/com/simbirsoft_mobile_revolution2017_v_team/mobilelibrary/repository/LibraryRepository.java
package com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.repository;

import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.domain.Book;
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.presenter.LibraryPresenter;
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.services.BaseResponse;
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.services.RestService;
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.services.RestServiceProvider;

import java.util.List;

import io.reactivex.Observable;

/**
 * Created by Olegka on 09.10.2017.
 */

public class LibraryRepository implements IRepository {

    @Override
    public Observable<List<Book>> getBooks() {
        return RestServiceProvider.getInstance().getRestService()
                .getBooks().map(BaseResponse::getData);
    }

    @Override
    public Observable<List<Book>> getFavouriteBooks() {
        return RestServiceProvider.getInstance().getRestService()
                .getFavouriteBooks(true).map(BaseResponse::getData);
    }

    @Override
    public Observable<Book> addBook(Book book) {
        return RestServiceProvider.getInstance().getRestService()
                .addBook(book).map(BaseResponse::getData);
    }

    @Override
    public Observable<Book> getBookById(String id){
        return RestServiceProvider.getInstance().getRestService()
                .getBook(id).map(BaseResponse::getData);
    }

}

File: app/src/main/java/com/simbirsoft_mobile_revolution2017_v_team/mobilelibrary/repository/LibraryRepository.kt
package com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.repository

import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.domain.BookK
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.services.BaseResponseK
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.services.RestServiceProviderK
import io.reactivex.Observable

/**
 * Created by user on 14.11.2017.
 */
class LibraryRepositoryK : IRepositoryK {
    override fun getBooks(): Observable<List<BookK>> {
        return RestServiceProviderK.getRestService()
                .getBooks().map(BaseResponseK<List<BookK>>::data)
    }

    override fun getFavouriteBooks(): Observable<List<BookK>> {
        return RestServiceProviderK.getRestService()
                .getFavouriteBooks(true).map(BaseResponseK<List<BookK>>::data)
    }

    override fun addBook(book: BookK): Observable<BookK> {
        return RestServiceProviderK.getRestService()
                .addBook(book).map(BaseResponseK<BookK>::data)
    }

    override fun getBookById(id: String): Observable<BookK> {
        return RestServiceProviderK.getRestService()
                .getBook(id).map(BaseResponseK<BookK>::data)
    }
}